docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+in the cache. 
+Instead of remembering the cached transactions, we could have had a valid/invalit bit per transaction within the cache. As transactions themselves do not
+store such information and we expect this scenario to be unlikely, instead of increasing the footprint of all transactions in the cache,
+we opted to keep a map of transaction signature if the transaction is in the cache, but is invalid. Alternatively, the cache could keep two lists, one for valid, and one for invalid transactions. 
+This modifies the following pieces of code as follows (this is just a prototype and does not include 
+some obvious sanity checks):
+
+```golang
+
+// New datastructure to keep track of peer failures
+type PeerFailure struct {
+  lastFailure time.Time
+  numFailures int8
+}
+
+type LRUTxCache struct {
+// Keeping track of invalid transactions within the cache ;
+  invalidCachedTx map[types.TxKey]bool
+}
+
+type CListMempool struct {
+ // ..existing fields
+  peerFailureMap map[nodeID]*PeerFailure
+
+}
+```
+
+>mempool/v0/clist_mempool.go#L239
+```golang
+
+if !mem.cache.Push(tx) { // if the transaction already exists in the cache
+		// Record a new sender for a tx we've already seen.
+		// Note it's possible a tx is still in the cache but no longer in the mempool
+		// (eg. after committing a block, txs are removed from mempool but not cache),
+		// so we only record the sender for txs still in the mempool.
+		if e, ok := mem.txsMap.Load(tx.Key()); ok {
+			memTx := e.(*clist.CElement).Value.(*mempoolTx)
+			memTx.senders.LoadOrStore(txInfo.SenderID, true)
+			// TODO: consider punishing peer for dups,
+			// its non-trivial since invalid txs can become valid,
+			// but they can spam the same tx with little cost to them atm.
+		}
+
+    // If transaction was invalid, we need to remember the peer information
+    if _, ok := mem.cache.invalidCachedTx.Load(tx.Key); ok {
+        mem.banPeer(peerID)  
+    }
+		return mempool.ErrTxInCache
+	}
+
+```
+
+
+```golang
+
+    func (mem* ClistMempool) banPeer(peerID NodeID) {
+        numFails := 0
+         // Check whether this peer has sent us transactions that fail
+        if val, ok := mem.peerFailureMap[peerID]; ok {
+         lastFailureT := val.lastFailure
+         numFails = val.numFails
+        // if the failure was recent enough, update the number of failures and 
+        // ban peer if applicable
+        if time.Since(lastFailureT) <= failureResetInterval {
+          if numFails == maxAllowedFailures - 1 {
+              // Send Ban request to p2p 
+            }
+          }
+        }
+        // Update the time of the last failure     
+        mem.peerFailureMap[peerID] = { time.Now(), numFailures + 1}
+    }
+```
+If transactions with `ResponseCheckTx.code > 1` are not deleted from the cache and we want to ban them on the first offence,
```

Comment:
Just to summarize our in person discussion which will also lead to a clarification in the doc. The debate we are having is whether we want the never-valid transactions in the cache or not. 
- Yes because this means we will never run checkTx on them again  and we know it will never be valid, so we may want it to stay there forever. 
- No because they are very rare, the peer is banned so they take up space in the cache when they don't occur that often